Make 'is_subset' work for ProductSet and FiniteSet  and Eq(Sets).simplify work

sidhu1012 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #19378

#### Brief description of what is fixed or changed
Earlier ProductSet.is_subset(Set) and Eq(Set).simplify() did't work .

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
    * Make 'is_subset' work for ProductSet
* core
    * Make class Eq with sets as arguments work with simplify().
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
This does not fix the issue with `Eq.simplify`.

sidhu1012 commented:
> This does not fix the issue with `Eq.simplify`.

Yes I'm working on that

sidhu1012 commented:
@oscarbenjamin I'll fix Eq.simplify() in different PR. I may be busy for a couple of days , so i think this PR can be merged till then. I'll change PR description.

sidhu1012 commented:
@oscarbenjamin Please review.

sidhu1012 commented:
It's ready for review!!
@oscarbenjamin @czgdp1807

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -311,8 +311,14 @@ def _eval_simplify(self, **kwargs):
+            else:
+                from sympy.sets import Set
+                if not isinstance(dif, Set):
+                    if dif.equals(0):  # XXX this is expensive
```
I don't think that sets should be a special case for sets here. There needs to be some kind of check here but it shouldn't be a `if not` negative check. The sets module should not be imported here in core.

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -507,6 +507,14 @@ def powerset(self):
         """
         return self._eval_powerset()
 
+    def equals(self, other):
```
I don't think sets should have a `.equals` method.

sidhu1012 commented:
@oscarbenjamin Ready for review!!

sidhu1012 commented:
@oscarbenjamin I think it's good to go.

sidhu1012 commented:
@oscarbenjamin Please have a look here.

sidhu1012 commented:
@oscarbenjamin I think it can be merged now.

oscarbenjamin commented on sympy/sets/handlers/issubset.py:
```diff
@@ -133,3 +133,9 @@ def is_subset_sets(a, b): # noqa:F811
+@dispatch(ProductSet, FiniteSet)  # type: ignore # noqa:F811
+def is_subset_sets(a_ps, b_fs): # noqa:F811
+    if not a_ps.is_finite_set:
+        return False
```
Can `is_finite_set` ever give `None`?

sidhu1012 replied:
I wasn't sure if ProductSet can be infinite or not, so I tackled it as a corner case.

oscarbenjamin replied:
You shouldn't use `not` if there is a possibility that `is_finite_set` can return `None`. Example:
```
In [12]: S = ConditionSet(x, Eq(x*y, 0), Reals)                                                                                   

In [13]: print(S.is_finite_set)                                                                                                   
None
```
In this example the set would be finite if y is equal to 1 but infinite if y is equal to 0.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -311,6 +311,9 @@ def _eval_simplify(self, **kwargs):
                 v = dif.n(2)
+            elif hasattr(dif, 'is_empty'):
+                if dif.is_empty:
+                    v = S.Zero
```
I'm not sure I understand what this does. Can you demonstrate?

sidhu1012 replied:
Earlier in case of sets, after `dif = lhs - rhs`  it tried to call `set.equal(0)`, i.e it tried to equate(considering `dif` = 0) empty set with 0 which was throwing error as `Sets` didn't have `.equal()` function , if that function is added, still comparing set with 0 would give error.

If `dif` is a set, empty or not it, would have `is_empty` attribute . Thus checking if `dif` is an empty set or not and proceeding and setting `v = S.Zero` accordingly.

In short in case of sets checking if `lhs - rhs`  is an `EmptySet` or not.

oscarbenjamin replied:
We should avoid computing `dif` altogether unless both sides are instances of Expr.

sidhu1012 replied:
But `dif` is playing nice with sets and is working as expected. Are you sure changing the entire method for sets is a good idea?

oscarbenjamin replied:
The logic in `Relational._eval_simplify` is clearly only intended for expressions. It is better to guard against any non-Expr inputs before calculating `dif`. Ideally we should restrict this block to number-type expressions but there isn't a clear way to do that.

For sets anyway it would probably be better to check equality using the subset relationship. Checking `(A - B)` is also not the same as checking `(B - A)` and it's possible that neither `A-B` or `B-A` is empty even though the sets are clearly unequal.

sidhu1012 replied:
Should I introduce handlers for specific cases like done for `is_subset` in sets?

oscarbenjamin replied:
I don't think that we need multiple dispatch in `Relational._eval_simplify` because both sides should always be of the same "type" in some sense. The only possible simplification here is that we can return True or False if we know that the two sides are or are not equal and that we can just determine using `is_subset(A, B)` and `is_subset(B, A)`.

sidhu1012 commented:
@oscarbenjamin Ready for review!

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -305,15 +305,17 @@ def _eval_simplify(self, **kwargs):
         r = self
         r = r.func(*[i.simplify(**kwargs) for i in r.args])
         if r.is_Relational:
+            if r.lhs.is_Set and r.rhs.is_Set:
+                if r.lhs.is_subset(r.rhs) and r.rhs.is_subset(r.lhs):
+                    return True
+                else:
+                    return False
             dif = r.lhs - r.rhs
```
The `is_subset` function does not always return True or False. You need to consider the indeterminate case e.g.:
```
In [1]: Eq({1}, {x})
Out[1]: {1} = {x}
```

oscarbenjamin replied:
Also the block below should only run if lhs and rhs are both Expr.

sidhu1012 commented:
@oscarbenjamin Please review now!

sidhu1012 commented:
Please review.

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -1597,6 +1597,16 @@ def test_issue_20089():
+    d = FiniteSet(1)
+    assert not b.is_subset(d)
+    assert Eq(b, c).simplify()
+    assert not Eq(a, c).simplify()
```
Tests that use `assert not f()` or `assert f()` should not be used. Almost any object returned by `f()` can be truthy so we should be more precise e.g.:
```
    assert Eq(b, c).simplify() is S.true
```
Likewise `None` can be falsey and we should not return `None` from `Eq.simplify` but this PR does:
```python
In [5]: x = Symbol('x')                                                                                                           

In [6]: e = Eq({1}, {x})                                                                                                          

In [7]: e                                                                                                                         
Out[7]: {1} = {x}

In [8]: print(simplify(e))                                                                                                        
None
```

sidhu1012 replied:
`Eq.simlpify` is returning `None` because of https://github.com/sympy/sympy/pull/20438#discussion_r534087321 . So wouldn't it be better to return `None` in indeterminate case ?
`None` can be changed to return `False` but that may not be correct.

sidhu1012 commented:
@oscarbenjamin  `   assert Eq(b, c).simplify() is S.true` assertion fails with `S.true`.

oscarbenjamin commented:
> `assert Eq(b, c).simplify() is S.true` assertion fails with `S.true`.

Exactly!

sidhu1012 commented:
@oscarbenjamin I have updated the test cases with `S.true` but what should be done of this https://github.com/sympy/sympy/pull/20438#discussion_r537256825.

oscarbenjamin commented:
Rather than asking me what to do it would be better if you took some time to review your own work.

Have you looked at how `Eq.simplify` works in general?

Have you read the docs for `Eq` or `simplify` to see if it says what kinds of objects should be returned?

Are your changes here consistent with the docstring and with the rest of the code for `Eq`?

Have you tested different examples from the existing code to see if you can understand what they do?

I'm happy to answer questions if you are confused about how things work in sympy in general. Please don't ask me to review your work though if you have not even spent a small amount of time checking it yourself.

sidhu1012 commented:
@oscarbenjamin Sorry, my bad, should have gone through docstring first. Figured it out though.

oscarbenjamin commented:
Closing and reopening to restart the tests

sidhu1012 commented:
@oscarbenjamin Why are all builds cancelled on Travis? Have we completely shifted to Actions?

sidhu1012 commented:
@oscarbenjamin Review pls.

sidhu1012 commented:
@oscarbenjamin Please review.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -105,6 +105,7 @@ class Basic(Printable, metaclass=ManagedProperties):
     is_Point = False
     is_MatAdd = False
     is_MatMul = False
+    is_Set = False
```
We already have too many of these `is_` attributes. It would be better not to add more. We can just do `isinstance(obj, Set)`.

sidhu1012 replied:
But earlier you said, we shouldn't import Set in core.

oscarbenjamin replied:
Oh, yes. I see what the problem is now.

The special casing of Set should be implemented by overriding `is_eq` as described here:
https://github.com/sympy/sympy/blob/25252ee4457461f7dfe568d1f495b4df6f7ca93d/sympy/core/relational.py#L1260

The code to override this can be in the sets module because it uses multiple dispatch.

sidhu1012 replied:
Done!

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -302,9 +302,19 @@ def equals(self, other, failing_expression=False):
         r = self
         r = r.func(*[i.simplify(**kwargs) for i in r.args])
         if r.is_Relational:
+            if r.lhs.is_Set and r.rhs.is_Set:
+                lhs = r.lhs.is_subset(r.rhs)
+                rhs = r.rhs.is_subset(r.lhs)
+                result = fuzzy_and([lhs, rhs])
+                if result is None:
+                    return self
+                return result
```
We can shortcut one of the `is_subset` calls if we do
```
result = fuzzy_and(a.is_subset(b) for a, b in [(lhs, rhs), (rhs, lhs)])
```
When this gives true or false it will return Python's `True` or `False` but it should return sympy's `S.true` or `S.false` because `Equality` is a symbolic `Boolean` so any simplification should also give a Boolean. Note that `S.true/false` are instances of `Basic` and have the associated methods like `subs` etc.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -575,7 +581,7 @@ def _eval_simplify(self, **kwargs):
                 if measure(enew) <= kwargs['ratio'] * measure(e):
                     e = enew
-            except ValueError:
+            except (ValueError, TypeError):
```
Why is TypeError being caught?

We should be very careful about catching generic exception types because it can hide bugs.

sidhu1012 replied:
When `Eq` is returned from `_eval_simplify` of `Relational` , `_eval_simplify` of `Equality` further tries to rewrite terms.

https://github.com/sympy/sympy/blob/d053c1d4c3ca5420851dcd9a30dfc0e7cee98a48/sympy/core/relational.py#L570

Here it tries to rewrite `Set` and throws `TypeError`.

oscarbenjamin replied:
Instead of catching the `TypeError` here there should be a check that both sides are Expr before the call to rewrite.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -302,9 +302,15 @@ def equals(self, other, failing_expression=False):
         r = r.func(*[i.simplify(**kwargs) for i in r.args])
         if r.is_Relational:
+            if not isinstance(r.lhs, Expr) or not isinstance(r.rhs, Expr):
+                result = _eval_is_eq(r.lhs, r.rhs)
```
This should use `is_eq` rather than `_eval_is_eq` but actually this shouldn't need to be here at all. The `r.func(...)` call above should be checking `is_eq`.

I think all we need to do here is avoid doing anything unless both sides are Expr.

oscarbenjamin commented on sympy/sets/handlers/comparison.py:
```diff
@@ -23,12 +23,6 @@ def _eval_is_eq(lhs, rhs): # noqa: F811
-
-@dispatch(FiniteSet, Interval) # type:ignore
-def _eval_is_eq(lhs, rhs): # noqa: F811
-    return False
```
Why is this removed?

sidhu1012 replied:
It was duplicate.This same is written twice in codebase.

oscarbenjamin commented on sympy/sets/handlers/comparison.py:
```diff
@@ -57,3 +51,29 @@ def _eval_is_eq(lhs, rhs): # noqa: F811
+    return tfn[False]
+
+
+@dispatch(FiniteSet, ProductSet) # type:ignore
```
I don't think it's necessary to define the dispatcher both ways round. The `is_eq` function will try flipping the arguments
https://github.com/sympy/sympy/blob/0bde5c951b52341e745d5410dba1a9a75d7acebf/sympy/core/relational.py#L1344-L1351

Even if it was necessary to do this both ways round it would clearly be better to use a function rather than duplicate code like this.

oscarbenjamin commented:
It might be better to override the general `_eval_is_eq(Set, Set)` to use `is_subset` rather than creating a specific `_eval_is_eq(ProductSet, FiniteSet)`.

Then if `is_subset` doesn't work for `ProductSet` and `FiniteSet` we can add handlers for that.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -302,9 +302,12 @@ def equals(self, other, failing_expression=False):
         r = r.func(*[i.simplify(**kwargs) for i in r.args])
         if r.is_Relational:
+            if not isinstance(r.lhs, Expr) or not isinstance(r.rhs, Expr):
+                return self
```
This should be `return r`

sidhu1012 replied:
Done!

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -1597,6 +1597,17 @@ def test_issue_20089():
+    c = FiniteSet((1, 1), (1, 2), (2, 1), (2, 2))
+    assert b.is_subset(c)
+    d = FiniteSet(1)
+    assert not b.is_subset(d)
```
These should be
```
assert b.is_subset(c) is True
```

sidhu1012 replied:
Done!

oscarbenjamin commented:
Okay, this looks good. Thanks for this.

Can you squash the commits and give a new commit message that summarises the final changes?

sidhu1012 commented:
> Okay, this looks good. Thanks for this.
> 
> Can you squash the commits and give a new commit message that summarises the final changes?

Done!

sidhu1012 commented:
@oscarbenjamin Can it be merged now?

sidhu1012 commented:
@oscarbenjamin Please review.

oscarbenjamin committed:
feat(sets): extend is_subset and fix equality of sets

Eq(SetA, SetB).simplify() is fixed so that it no longer raises.
Handlers are added so that

  Eq(SetA, SetB)

can resolve itself through

  is_eq(SetA, SetB)

which in turn calls

  SetA.is_subset(Setb) and SetB.is_subset(SetA)

Some new handlers for is_subset to cover more cases.

Co-authored-by: Sudeep Sidhu <sudeepmanilsidhu@gmail.com>

oscarbenjamin commented:
I've squashed the commits and force pushed onto here. This can be merged if the tests pass.
